mcs/class/System/ReferenceSources/Win32Exception.cs

```diff
@@ -1,15 +1,1752 @@
+
+using Microsoft.Win32;
+using System.Collections.Generic;
 using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading;
 
 namespace System.ComponentModel
 {
 	partial class Win32Exception
 	{
-		[MethodImplAttribute (MethodImplOptions.InternalCall)]
-		internal static extern string W32ErrorMessage (int error_code);
+		const int FORMAT_MESSAGE_IGNORE_INSERTS = 0x00000200;
+		const int FORMAT_MESSAGE_FROM_SYSTEM    = 0x00001000;
+
+		[DllImport ("api-ms-win-core-localization-l1-2-0.dll", CharSet = CharSet.Unicode,
+			EntryPoint = "FormatMessageW", SetLastError = true, BestFitMapping = true)]
+		private static extern int FormatMessage(int dwFlags, IntPtr lpSource, uint dwMessageId, int dwLanguageId,
+			[Out] StringBuilder lpBuffer, int nSize, IntPtr[] arguments);
+
+		internal static string GetErrorMessage (int error)
+		{
+			if (Environment.OSVersion.Platform != PlatformID.Unix) {
+				StringBuilder sb = new StringBuilder (256);
+
+				int result = FormatMessage (FORMAT_MESSAGE_IGNORE_INSERTS | FORMAT_MESSAGE_FROM_SYSTEM,
+					IntPtr.Zero, (uint)error, 0, sb, sb.Capacity, null);
+
+				if (result == 0)
+					return "Error looking up error string";
+
+				return sb.ToString ();
+			}
+
+			string message;
+			if (!Messages.Value.TryGetValue (error, out message))
+				return string.Format ("mono-io-layer-error ({0})", error);
+
+			return message;
+		}
+
+		static Lazy<Dictionary<int, string>> Messages =
```

Comment:
how about the performance? switch has complexity O(n) while the dictionary has complexity O(1). And creating this dictionary takes 3ms on my machine.